network/src/noise_wrapper/stream.rs

```diff
@@ -557,163 +426,204 @@ where
+    /// A write buffer, used for both a plaintext to send, and then its encrypted version.
+    write_buffer: [u8; noise::MAX_SIZE_NOISE_MSG],
+}
 
-            // <- e, ee, se, s, es
-            self.send().await?;
-            self.flush().await?;
+impl NoiseBuffers {
+    fn new() -> Self {
+        Self {
+            read_buffer: [0; noise::MAX_SIZE_NOISE_MSG],
+            write_buffer: [0; noise::MAX_SIZE_NOISE_MSG],
         }
-
-        self.finish()
     }
+}
 
-    /// Send handshake message to remote.
-    async fn send(&mut self) -> io::Result<()> {
-        poll_fn(|context| self.0.poll_write(context, &[]))
-            .await
-            .map(|_| ())
+/// Hand written Debug implementation in order to omit the printing of huge buffers of data
+impl ::std::fmt::Debug for NoiseBuffers {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        f.debug_struct("NoiseBuffers").finish()
     }
+}
+
+//
+// Helpers for writing to and reading from a socket
+// ------------------------------------------------
+//
+
+/// Write an offset of a buffer to a socket, only returns Ready once done.
+fn poll_write_all<TSocket>(
+    mut context: &mut Context,
+    mut socket: Pin<&mut TSocket>,
+    buf: &[u8],
+    offset: &mut usize,
+) -> Poll<io::Result<()>>
+where
+    TSocket: AsyncWrite,
+{
+    loop {
+        let n = ready!(socket.as_mut().poll_write(&mut context, &buf[*offset..]))?;
+        trace!("poll_write_all: wrote {}/{} bytes", *offset + n, buf.len());
+        if n == 0 {
+            return Poll::Ready(Err(io::ErrorKind::WriteZero.into()));
+        }
+        *offset += n;
+        assert!(*offset <= buf.len());
 
-    /// Flush handshake message to remote.
-    async fn flush(&mut self) -> io::Result<()> {
-        poll_fn(|context| self.0.poll_flush(context)).await
+        if *offset == buf.len() {
+            return Poll::Ready(Ok(()));
+        }
     }
+}
 
-    /// Receive handshake message from remote.
-    async fn receive(&mut self) -> io::Result<()> {
-        poll_fn(|context| self.0.poll_read(context, &mut []))
-            .await
-            .map(|_| ())
+/// Read a u16 frame length from `socket`.
+///
+/// Can result in the following output:
+/// 1) Ok(None) => EOF; remote graceful shutdown
+/// 2) Err(UnexpectedEOF) => read 1 byte then hit EOF; remote died
+/// 3) Ok(Some(n)) => new frame of length n
+fn poll_read_u16frame_len<TSocket>(
+    context: &mut Context,
+    socket: Pin<&mut TSocket>,
+    buf: &mut [u8; 2],
+    offset: &mut usize,
+) -> Poll<io::Result<Option<u16>>>
+where
+    TSocket: AsyncRead,
+{
+    match ready!(poll_read_exact(context, socket, buf, offset)) {
+        Ok(()) => Poll::Ready(Ok(Some(u16::from_be_bytes(*buf)))),
+        Err(e) => {
+            if *offset == 0 && e.kind() == io::ErrorKind::UnexpectedEof {
+                return Poll::Ready(Ok(None));
+            }
+            Poll::Ready(Err(e))
+        }
```

Comment:
let res = match ready!(poll_read_exact(context, socket, buf, offset)) {
      Ok(()) => Ok(Some(u16::from_be_bytes(*buf)))),
      Err(io::ErrorKind::UnexpectedEof) if *offset == 0 =>  Ok(None),
      Err(e) => Err(e),
   };
   Poll::Ready(res)
Would something like this help?

Reply:
I'm actually just copying this code, and it works, so I'm not touching it : D

Reply:
The answer is usually to put the change into a single commit, so that it's independently reversible.